File: example_person.py
# -*- coding:UTF-8 -*-

import sys
import os
import proto_pb2.Person_pb2 as GeneralProto
import proto_buf.Person_buf_read as PersonRead
import proto_buf.Person_buf_write as PersonWrite
import simplejson


if __name__ == "__main__":
  if not os.path.exists(os.path.join(os.getcwd(), "data")):
    os.mkdir("data")
  
  protofile = os.path.join(os.getcwd(), "data/Person_test.proto")
  version = 20191001
  writer = PersonWrite.GeneralProtoWriter(protofile, GeneralProto.Person, version)

  person1 = GeneralProto.Person()
  person1.id = 100000
  person1.name = "zhangsan"
  person1.age = 20

  person1.email.append("123456@qq.com")
  person1.email.append("234567@qq.com")
  phone1 = person1.phone.add()
  phone2 = person1.phone.add()
  phone1.number = "987654"
  phone1.type = GeneralProto.PhoneType.MOBILE
  phone2.number = "876543"
  phone2.type = GeneralProto.PhoneType.HOME

  addr = person1.address
  addr.country = "china"
  addr.detail = "beijing"

  writer.writeFrameData_general(person1)

  person2 = '{"name":"lisi","age":22,"id":200000,"email":["345678@qq.com","456789@qq.com"],"phone":[{"type":1,"number":"765432"},{"type":2,"number":"654321"}],"address":{"country":"china","detail":"nanjing"}}'

  writer.writeFrameData_json(person2)
  writer.stopWriter()

  reader = PersonRead.GeneralProtoReader(protofile, GeneralProto.Person)
  version_read = reader.getVersion()

  reader.setFrameIndex(0)
  print(reader.getFrameData_general("id"))
  print(reader.getFrameData_general("name"))
  print(reader.getFrameData_general("age"))
  print(reader.getFrameData_general("email"))
  print(reader.getFrameData_general("phone"))
  print(reader.getFrameData_general("address.country"))
  print(reader.getFrameData_general("address.detail"))
  
  reader.setFrameIndex(1)
  json = reader.getFrameData_json()
  print(json)

File: c++/example_person.cpp
#include <iostream>
#include "General_buf_read.h"
#include "General_buf_write.h"
#include "Person.pb.h" 

namespace PersonProto {
  class Person;
};

namespace GeneralBuf {
  template <typename T>
  class GeneralProtoWriter;
  typedef GeneralProtoWriter<PersonProto::Person> PersonProtoWriter;
  template <typename T>
  class GeneralProtoReader;
  typedef GeneralProtoReader<PersonProto::Person> PersonProtoReader;
};

int main(int argc, char const *argv[])
{
  GeneralBuf::PersonProtoWriter writer;
  std::string filename = "Person_test.proto";
  int64_t version = 20191001;
  writer.startWriter(filename, version);

  PersonProto::Person *person1 = new PersonProto::Person();
  person1->set_id(100000);
  person1->set_name("zhangsan");
  person1->set_age(20);

  person1->add_email("123456@qq.com");
  person1->add_email("234567@qq.com");
  PersonProto::PhoneNumber *phone1 = person1->add_phone();
  PersonProto::PhoneNumber *phone2 = person1->add_phone();
  phone1->set_number("987654");
  phone1->set_type(PersonProto::PhoneType::MOBILE);
  phone2->set_number("876543");
  phone2->set_type(PersonProto::PhoneType::HOME);

  PersonProto::Address *addr = person1->mutable_address();
  addr->set_country("china");
  addr->set_detail("beijing");

  writer.write(person1);
  writer.stopWriter();

  GeneralBuf::PersonProtoReader reader;
  reader.startReader(filename);

  std::cout << "version : " << reader.getVersion() << "\n";
  std::cout << "cnt : " << reader.getFrameCount() << "\n";
  reader.setFrameIndex(0);
  PersonProto::Person *person = new PersonProto::Person();
  reader.read(person);
  person->PrintDebugString();

  return 0;
}
